### PR TITLE
Backport Spark adaptor to 2.5

### PR DESCRIPTION
**Pull request by jyuanca**
backport a series of commits from master to branch-2.5 to make spark adaptor work

### CODE REVIEW

**Comment by jyuanca** [comment]
@dgomezferro Yes, they are needed.

File: db-client/src/main/java/com/splicemachine/db/client/am/ColumnMetaData.java [hunk 1]
```diff
@@ -329,6 +329,7 @@ public int getColumnDisplaySize(int column) throws SQLException {
             case Types.TIMESTAMP:
                 return 29;
             case Types.JAVA_OBJECT:
+            case Types.ARRAY:
```
Comment (wtang16) [inline db-client/src/main/java/com/splicemachine/db/client/am/ColumnMetaData.java]: ARRAY data type is not in 2.5. I am not sure if we need this case statement.

File: db-client/src/main/java/com/splicemachine/db/client/am/Cursor.java [hunk 2]
```diff
@@ -1062,7 +1062,8 @@ final String getString(int column) throws SqlException {
                 return (maxFieldSize_ == 0) ? tempString :
                         tempString.substring(0, java.lang.Math.min(maxFieldSize_, tempString.length()));
             case java.sql.Types.JAVA_OBJECT:
-                Object obj = get_UDT( column );
+            case java.sql.Types.ARRAY:
```
Comment (wtang16) [inline db-client/src/main/java/com/splicemachine/db/client/am/Cursor.java]: ARRAY is not in 2.5.

File: db-client/src/main/java/com/splicemachine/db/client/am/PreparedStatement.java [hunk 3]
```diff
@@ -975,7 +975,16 @@ public void setBytesX(int parameterIndex, byte[] x) throws SqlException {
...
+    public void setArrayX(int parameterIndex, Array x) throws SqlException {
+        parameterMetaData_.clientParamtertype_[parameterIndex - 1] = java.sql.Types.ARRAY;
+        if (x == null) {
+            setNullX(parameterIndex, java.sql.Types.ARRAY);
+            return;
+        }
+        setInput(parameterIndex, x);
+    }
+
     /**
```
Comment (wtang16) [inline db-client/src/main/java/com/splicemachine/db/client/am/PreparedStatement.java]: ARRY is out of 2.5.

**Pull request closed without merging**